serviceaccounts.rb

```diff
@@ -18,20 +21,39 @@ def initialize(project, service_account = nil, path = nil)
     end
   end
 
+  def existing_file_account(path)
+    if File.exists?(path)
+      begin
+        return JSON.parse(File.read(path))["client_email"]
+      rescue JSON::ParserError => e
+        return nil
+      end
+    end
+    return nil
+  end
+
   def run()
-    ENV["GOOGLE_APPLICATION_CREDENTIALS"] = @path
     common = Common.new
+    ENV["GOOGLE_APPLICATION_CREDENTIALS"] = @path
+    service_account = @service_account
+    if not service_account
+      service_account = "#{@project}@appspot.gserviceaccount.com"
+    end
+    if service_account == existing_file_account(@path)
+      # Don't generate another key if this account is already active. This can
+      # happen for nested service account contexts, for example.
+      common.status "Attaching to existing keyfile @ #{@path}"
+      yield
+      return
+    end
+
     if @project == "all-of-us-workbench-test" and not @service_account
```

Comment:
Because of the "not @service_account" I don't think this runs anymore.

If it did run, we should probably handle the possibility that existing_file_account(@path) doesn't match.

The goal of this logic was to make it so that specifically for the test SA, we copy credentials from GCS rather than making new ones... the idea being that we don't want all of our developers trying to create test SA keys simultaneously or we will run out.